File: cpp/1300.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findBestValue(vector<int>& arr, int target) {
        sort(arr.begin(), arr.end());  // sort
        int l = 0, r = arr.size()-1;
        // int pos =  (l + r) / 2;
        int r_val = arr[r], l_val = arr[l];
        int res = (r_val + l_val) / 2, mid = (l+r)/2;
        while (l < r) {
            /* code */
                        
        }
        return res;
    }
};

File: python/1300.py
class Solution:
    def findBestValue(self, arr: List[int], target: int) -> int:
        def _sum(arr, m):
            __sum = 0
            for val in arr:
                __sum += val if val <= m else m
            return __sum

        l, r = 0, max(arr)
        while l < r:
            mid = (l + r)//2
            temp = _sum(arr, mid)
            print(l, r, mid, temp)
            if temp > target:
                r = mid
            elif temp < target:
                l = mid + 1
            else:
                return mid

        return l-1 if abs(_sum(arr, l) - target) >= abs(_sum(arr, l-1) - target) else l